### PR TITLE
Make responders into GenServers

### PR DESCRIPTION
**Pull request by scrogson**
This change makes responders into isolated `GenServer`s allowing responders to maintain/update state.

**TODO**

- [x] Update docs
- [x] More tests

### CODE REVIEW

File: lib/hedwig/responder.ex [hunk 1]
```diff
@@ -193,48 +176,74 @@ defmodule Hedwig.Responder do
...
       end
 
-      defp install_hearer({regex, fun}, _robot, opts) do
-        Task.async(fn ->
-          {regex, __MODULE__, fun, Enum.into(opts, %{})}
-        end)
+      @doc false
+      def usage(name) do
+        @usage
+        |> Enum.map(&String.strip/1)
```
Comment (knewter) [inline lib/hedwig/responder.ex]: What about `&(String.strip(&1) |> String.replace("hedwig", name))`...or even:

```elixir
def usage(name) do
  import String
  @usage
  |> Enum.map(&(strip(&1) |> replace("hedwig", name))
```

Or is this worse? :)

Comment (knewter) [inline lib/hedwig/responder.ex]: (I never knew how much I wanted automatic currying til i had it)

File: lib/hedwig/robot.ex [hunk 2]
```diff
@@ -69,51 +69,57 @@ defmodule Hedwig.Robot do
...
       def __adapter__, do: @adapter
 
       def init({robot, opts}) do
-        opts = Keyword.merge(robot.config, opts)
-        {:ok, adapter} = @adapter.start_link(robot, opts)
-
-        {aka, opts}   = Keyword.pop(opts, :aka)
-        {name, opts}  = Keyword.pop(opts, :name)
-        responders    = Keyword.get(opts, :responders, [])
+        opts = robot.config(opts)
```
Comment (knewter) [inline lib/hedwig/robot.ex]: ever considered `with` here?  I'm obviously bikeshedding

File: lib/hedwig/robot.ex [hunk 3]
```diff
@@ -69,51 +69,57 @@ defmodule Hedwig.Robot do
...
-        #{inspect __MODULE__}.handle_in/2 default handler invoked.
-        """
-        {:ok, state}
+        case msg do
+          %Hedwig.Message{} ->
+            {:dispatch, msg, state}
+          _ ->
+            {:noreply, state}
+        end
+      end
```
Comment (knewter) [inline lib/hedwig/robot.ex]:
```elixir
def handle_in(%Hedwig.Message{}=msg, state), do: {:dispatch, msg, state}
def handle_in(_, state), do: {:noreply, state}
```
?

File: lib/hedwig/robot.ex [hunk 4]
```diff
@@ -156,22 +162,36 @@ defmodule Hedwig.Robot do
         {:noreply, state}
       end
 
-      def handle_cast(%Hedwig.Message{} = msg, %{responders: responders} = state) do
-        Hedwig.Responder.run(%{msg | robot: %{state | responders: []}}, responders)
-        {:noreply, state}
-      end
+      def handle_cast({:handle_in, msg}, %{responder_sup: sup} = state) do
+        case __MODULE__.handle_in(msg, state) do
```
Comment (knewter) [inline lib/hedwig/robot.ex]: any reason you can't omit `__MODULE__` here?

Comment (scrogson) [inline lib/hedwig/robot.ex]: good point.

File: lib/hedwig/robot.ex (hunk 3 above)
Comment (scrogson) [inline lib/hedwig/robot.ex]: Yeah, I do like this a bit better. Also...could add these to the Robot generator.

File: lib/hedwig/robot.ex (hunk 2 above)
Comment (scrogson) [inline lib/hedwig/robot.ex]: :troll:

**Pull request merged**